### PR TITLE
Simplify data types in interfaces and implementations

### PR DESCRIPTION
**Pull request by garth-wells**
Use more `std::vector` and `span`. Make performance implications clearer in the numerous places.

### CODE REVIEW

File: cpp/dolfinx/fem/assemble_matrix_impl.h [hunk 1]
```diff
@@ -428,8 +428,10 @@ void assemble_interior_facets(
     {
       for (int j = 0; j < gdim; ++j)
       {
-        coordinate_dofs(i, j) = x_g(x_dofs0[i], j);
-        coordinate_dofs(i + num_dofs_g, j) = x_g(x_dofs1[i], j);
+        // coordinate_dofs(i, j) = x_g(x_dofs0[i], j);
+        coordinate_dofs[i * gdim + j] = x_g(x_dofs0[i], j);
+        // coordinate_dofs(i + num_dofs_g, j) = x_g(x_dofs1[i], j);
```
Comment (michalhabera) [inline cpp/dolfinx/fem/assemble_matrix_impl.h:433]: Leftover comments.

File: cpp/dolfinx/fem/assemble_scalar_impl.h [hunk 2]
```diff
@@ -303,8 +305,10 @@ T assemble_interior_facets(
     {
       for (int j = 0; j < gdim; ++j)
       {
-        coordinate_dofs(i, j) = x_g(x_dofs0[i], j);
-        coordinate_dofs(i + num_dofs_g, j) = x_g(x_dofs1[i], j);
+        // coordinate_dofs(i, j) = x_g(x_dofs0[i], j);
```
Comment (michalhabera) [inline cpp/dolfinx/fem/assemble_scalar_impl.h:308]: Same here.

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 3]
```diff
@@ -196,10 +200,10 @@ void _lift_bc_cells(
         if (bc_markers1[jj])
         {
           const T bc = bc_values1[jj];
-          if (!x0.empty())
-            be -= Ae.col(bs1 * j + k) * scale * (bc - x0[jj]);
-          else
-            be -= Ae.col(bs1 * j + k) * scale * bc;
+          const T _x0 = x0.empty() ? 0.0 : x0[jj];
+          // be -= Ae.col(bs1 * j + k) * scale * (bc - _x0);
```
Comment (michalhabera) [inline cpp/dolfinx/fem/assemble_vector_impl.h:204]: Commented code also here, probably in most assembly functions?

File: cpp/dolfinx/fem/assemble_matrix_impl.h (hunk 1 above)
Comment (garth-wells) [reply cpp/dolfinx/fem/assemble_matrix_impl.h:433]: I left it there deliberately to explain the operation.

File: cpp/dolfinx/fem/assemble_vector_impl.h (hunk 3 above)
Comment (garth-wells) [reply cpp/dolfinx/fem/assemble_vector_impl.h:204]: Yes, for readability to the reader doesn't have to reverse engineer the operaion.

File: cpp/dolfinx/fem/assemble_matrix_impl.h (hunk 1 above)
Comment (michalhabera) [reply cpp/dolfinx/fem/assemble_matrix_impl.h:433]: Well, seems quite obvious (just flattened array access). I would not read it as a hint, as this is a valid commented code I would treat it as some leftover.

File: cpp/dolfinx/fem/assemble_vector_impl.h (hunk 3 above)
Comment (garth-wells) [reply cpp/dolfinx/fem/assemble_vector_impl.h:204]: I've left this one because it's helpful to summarise the 'math' operation.

**Pull request merged**